plugins/cpp_reparse/service/src/typespecialmembers.cpp

```diff
@@ -0,0 +1,1502 @@
+#include <algorithm>
+#include <iterator>
+#include <sstream>
+#include <utility>
+
+#include <boost/algorithm/string/regex.hpp>
+
+#include <clang/AST/RecursiveASTVisitor.h>
+#include <clang/Basic/Diagnostic.h>
+
+#include <util/logutil.h>
+
+#include <model/file.h>
+#include <model/file-odb.hxx>
+
+#include <cppparser/filelocutil.h>
+#include <service/reparser.h>
+
+#include "typespecialmembers.h"
+
+namespace
+{
+
+using namespace cc;
+using namespace cc::service::language;
+using namespace cc::service::reparse;
+using namespace clang;
+using namespace llvm;
+
+typedef odb::query<model::File> FileQuery;
+
+typedef std::map<void*, std::string> ASTNodeToIdentifierMapTy;
+
+template <typename It, typename ElemFn>
+void printSeparatedList(std::ostream& os_,
+                        const std::string& separator_,
+                        It begin_, It end_, ElemFn fn_)
+{
+  while (begin_ != end_)
+  {
+    fn_(*begin_);
+
+    ++begin_;
+    if (begin_ != end_)
+      os_ << separator_;
+  }
+}
+
+void printType(std::ostream& os_, const QualType& type_)
+{
+  auto printQualifiers = [&os_](const QualType& type)
+  {
+    if (type.isConstQualified())
+      os_ << "const ";
+    if (type.isVolatileQualified())
+      os_ << "volatile ";
+    if (type.isRestrictQualified())
+      os_ << "restrict ";
+  };
+
+  if (type_->isReferenceType())
+    printQualifiers(cast<ReferenceType>(type_)->getPointeeTypeAsWritten());
+  else
+    printQualifiers(type_);
+
+  os_ << type_.getBaseTypeIdentifier()->getName().str();
+
+  if (type_->isLValueReferenceType())
+    os_ << "&";
+  else if (type_->isRValueReferenceType())
+    os_ << "&&";
+}
+
+void printParameters(std::ostream& os_,
+                     ASTNodeToIdentifierMapTy& argMap_,
+                     FunctionDecl::param_iterator begin_,
+                     FunctionDecl::param_iterator end_)
+{
+  // Set up a mapping for the ParmVarDecls to a dummy name. In the AST,
+  // the implicit parameters might not have a name.
+  assert(argMap_.empty() && "printParameters sets up argument-name bindings"
+                            "but was given a non-empty map.");
+  {
+    size_t argNum = 1;
+    FunctionDecl::param_iterator begin = begin_;
+
+    while (begin != end_)
+    {
+      ParmVarDecl& var = **begin;
+
+      if (!var.getName().empty())
+        argMap_[&var] = var.getName().str();
+      else
+      {
+        argMap_[&var] = "arg" + std::to_string(argNum);
```

Comment:
Compiler generated special member functions **do not** have a variable name, because the Clang AST is interconnected on the memory address level for the `DeclRefExpr`s that actually refer to the implicit parameter. See the following Clang AST excerpt:

![clangastdump](https://user-images.githubusercontent.com/1969470/47957411-7207f980-dfb5-11e8-8eef-d03debba6fce.png)

See the `DeclRefExpr`'s row, after the `ParmVarDecl` and it's memory address, the variable's name is an empty string between the two cyan-coloured apostrophes.